pkg/cluster/mpi/statusupdate.go

```diff
@@ -91,28 +89,43 @@ func (c statusUpdateComponent) Reconcile(ctx *core.Context) (ctrl.Result, error)
 
 	if modified {
 		err = ctx.Client.Status().Update(ctx, cr)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("cannot update cluster status: %w", err)
+		}
 	}
 
-	// TODO: scale down workers after job is "Complete" or "Failed"
-	// 	this component focuses on updating the "status" so the downscaling
-	// 	logic probably belongs elsewhere
-
-	return ctrl.Result{}, err
+	return ctrl.Result{}, nil
 }
 
-func getActivePodCnt(ctx *core.Context, cr *dcv1alpha1.MPICluster) (int, error) {
-	var ep corev1.Endpoints
-	objKey := client.ObjectKey{
-		Name:      serviceName(cr),
-		Namespace: cr.Namespace,
+func (c statusUpdateComponent) Finalize(ctx *core.Context) (ctrl.Result, bool, error) {
+	cr := objToMPICluster(ctx.Object)
+
+	if cr.Status.ClusterStatus != StoppingStatus {
+		cr.Status.ClusterStatus = StoppingStatus
+		err := ctx.Client.Status().Update(ctx, cr)
+		if err != nil {
+			return ctrl.Result{}, false, fmt.Errorf("cannot update cluster status: %w", err)
+		}
 	}
 
-	if err := ctx.Client.Get(ctx, objKey, &ep); err != nil && !apierrors.IsNotFound(err) {
-		return 0, fmt.Errorf("cannot fetch endpoints: %w", err)
+	pods, err := getPods(ctx, cr)
```

Comment:
What is the point of this check? I think if the StatefulSet is gone, then the Pods can be assumed to be gone.

Reply:
You mean, check on the stateful set, not the pods? Yeah, perhaps. It's probably a more correct way of getting a result, but on the other hand I already have a `getPods` function handy.

Reply:
I would say: If the statefulset resource finalizer is done, then the statefulset should not exist, and we can assume all owned resources of the statefulset (which includes pods) do not exist.

Reply:
Turns out, we need to check specifically for pods here (not for the stateful set), because the stateful set won't wait for pods if deleted on the foreground.